File: version_c/colas.c
#include <stdio.h>
#include <stdlib.h>
#include "colas.h"

// Prepara la cola
void inicializar_cola_estatica(ColaEstatica *cola) {
    cola->frente = 0;
    cola->final = 0;
    cola->tamano = 0;
}

// Agrega al final
void enqueue_estatica(ColaEstatica *cola, int valor) {
    if (cola->tamano == CAPACIDAD_COLA) {
        printf("Cola llena\n");
        return;
    }
    cola->datos[cola->final] = valor;
    cola->final = (cola->final + 1) % CAPACIDAD_COLA;
    cola->tamano++;
    printf("Elemento agregado\n");
}

// Elimina del frente
void dequeue_estatica(ColaEstatica *cola) {
    int valor;
    if (cola->tamano == 0) {
        printf("Cola vacia\n");
        return;
    }
    valor = cola->datos[cola->frente];
    printf("Elemento a eliminar: %d\n", valor);
    printf("Eliminado: %d\n", valor);
    cola->frente = (cola->frente + 1) % CAPACIDAD_COLA;
    cola->tamano--;
}

// Muestra la cola
void mostrar_cola_estatica(ColaEstatica *cola) {
    int i;
    int indice = cola->frente;
    printf("Cola: [");
    for (i = 0; i < cola->tamano; i++) {
        printf("%d", cola->datos[indice]);
        if (i < cola->tamano - 1) {
            printf(" ");
        }
        indice = (indice + 1) % CAPACIDAD_COLA;
    }
    printf("]\n");
}

// Prepara la cola
void inicializar_cola_dinamica(ColaDinamica *cola) {
    cola->frente = NULL;
    cola->final = NULL;
}

// Agrega al final
void enqueue_dinamica(ColaDinamica *cola, int valor) {
    NodoCola *nuevo = (NodoCola *) malloc(sizeof(NodoCola));
    nuevo->valor = valor;
    nuevo->siguiente = NULL;
    if (cola->final == NULL) {
        cola->frente = nuevo;
        cola->final = nuevo;
    } else {
        cola->final->siguiente = nuevo;
        cola->final = nuevo;
    }
    printf("Elemento agregado\n");
}

// Elimina del frente
void dequeue_dinamica(ColaDinamica *cola) {
    NodoCola *temporal;
    if (cola->frente == NULL) {
        printf("Cola vacia\n");
        return;
    }
    temporal = cola->frente;
    printf("Elemento a eliminar: %d\n", temporal->valor);
    printf("Eliminado: %d\n", temporal->valor);
    cola->frente = temporal->siguiente;
    if (cola->frente == NULL) {
        cola->final = NULL;
    }
    free(temporal);
}

// Muestra la cola
void mostrar_cola_dinamica(ColaDinamica *cola) {
    NodoCola *actual = cola->frente;
    printf("Cola: [");
    while (actual != NULL) {
        printf("%d", actual->valor);
        if (actual->siguiente != NULL) {
            printf(" ");
        }
        actual = actual->siguiente;
    }
    printf("]\n");
}

// Libera memoria
void liberar_cola_dinamica(ColaDinamica *cola) {
    while (cola->frente != NULL) {
        NodoCola *temporal = cola->frente;
        cola->frente = cola->frente->siguiente;
        free(temporal);
    }
    cola->final = NULL;
}

File: colas.py
from nodos import Nodo


class ColaEstatica:
    def __init__(self, capacidad=5):
        self.capacidad = capacidad
        self.datos = [None] * capacidad
        self.frente = 0
        self.final = 0
        self.tamano = 0

    # Agrega al final
    def enqueue(self, valor):
        if self.tamano == self.capacidad:
            print("Cola llena.")
            return
        self.datos[self.final] = valor
        self.final = (self.final + 1) % self.capacidad
        self.tamano += 1
        print("Elemento agregado.")

    # Elimina del frente
    def dequeue(self):
        if self.tamano == 0:
            print("Cola vacia.")
            return
        valor = self.datos[self.frente]
        print(f"Elemento a eliminar: {valor}")
        self.datos[self.frente] = None
        self.frente = (self.frente + 1) % self.capacidad
        self.tamano -= 1
        print(f"Eliminado: {valor}")

    # Muestra la cola
    def mostrar(self):
        resultado = []
        indice = self.frente
        for _ in range(self.tamano):
            resultado.append(self.datos[indice])
            indice = (indice + 1) % self.capacidad
        print(f"Cola: {resultado}")


class ColaDinamica:
    def __init__(self):
        self.frente = None
        self.final = None

    # Recorre los nodos
    def recorrer(self):
        actual = self.frente
        while actual:
            yield actual
            actual = actual.siguiente

    # Agrega al final
    def enqueue(self, valor):
        nuevo = Nodo(valor)
        if self.final is None:
            self.frente = nuevo
            self.final = nuevo
        else:
            self.final.siguiente = nuevo
            self.final = nuevo
        print("Elemento agregado.")

    # Elimina del frente
    def dequeue(self):
        if self.frente is None:
            print("Cola vacia.")
            return
        print(f"Elemento a eliminar: {self.frente.valor}")
        valor = self.frente.valor
        self.frente = self.frente.siguiente
        if self.frente is None:
            self.final = None
        print(f"Eliminado: {valor}")

    # Muestra la cola
    def mostrar(self):
        datos = []
        for actual in self.recorrer():
            datos.append(actual.valor)
        print(f"Cola: {datos}")
